Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,588 @@
+-- |
+-- Module:      Math.NumberTheory.DirichletCharacters
+-- Copyright:   (c) 2018 Bhavik Mehta
+-- Licence:     MIT
+-- Maintainer:  Bhavik Mehta <bhavikmehta8@gmail.com>
+--
+-- Implementation and enumeration of Dirichlet characters.
+--
+
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE PatternSynonyms            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE ViewPatterns               #-}
+
+module Math.NumberTheory.DirichletCharacters
+  (
+  -- * Roots of unity
+    RootOfUnity
+  -- ** Conversions
+  , toRootOfUnity
+  , fromRootOfUnity
+  , toComplex
+  -- * An absorbing semigroup
+  , OrZero, pattern Zero, pattern NonZero
+  , asNumber
+  -- * Dirichlet characters
+  , DirichletCharacter
+  -- ** Construction
+  , indexToChar
+  , indicesToChars
+  , characterNumber
+  , allChars
+  -- ** Evaluation
+  , eval
+  , evalGeneral
+  , toFunction
+  , evalAll
+  -- ** Special Dirichlet characters
+  , principalChar
+  , isPrincipal
+  , orderChar
+  -- ** Real Dirichlet characters
+  , RealCharacter
+  , isRealCharacter
+  , getRealChar
+  , toRealFunction
+  , jacobiCharacter
+  -- ** Primitive characters
+  , PrimitiveCharacter
+  , isPrimitive
+  , getPrimitiveChar
+  , induced
+  , makePrimitive
+  , WithNat(..)
+  -- * Debugging
+  , validChar
+  ) where
+
+import Control.Applicative                                 (liftA2)
+import Data.Bits                                           (Bits(..))
+import Data.Complex                                        (Complex(..), cis)
+import Data.Foldable                                       (for_)
+import Data.Functor.Identity                               (Identity(..))
+import Data.List                                           (mapAccumL, foldl', sort, find)
+import Data.Maybe                                          (mapMaybe)
+#if MIN_VERSION_base(4,12,0)
+import Data.Monoid                                         (Ap(..))
+#endif
+import Data.Proxy                                          (Proxy(..))
+import Data.Ratio                                          (Rational, (%), numerator, denominator)
```

Comment:
I'm getting a warning that `The import of ‘Rational’ from module ‘Data.Ratio’ is redundant`.